org.eclipse.jdt.junit.core/src/org/eclipse/jdt/internal/junit/model/TestSuiteElement.java

```diff
@@ -47,7 +47,7 @@ public String getSuiteTypeName() {
 
 	@Override
 	public ITestElement[] getChildren() {
-		return fChildren.toArray(new ITestElement[fChildren.size()]);
+		return fChildren.stream().filter(e -> !isSingleDynamicTest(e)).toArray(ITestElement[]::new);
```

Comment:
I haven't debugged that yet. 
How often is this called? I fear the filter may slowdown execution if this method is called on every view update and there are ~10.000 tests.

Can you please try to run JDT compiler regression tests (they run parametrized by JLS) and check if this change breaks anything:
org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/TestAll.java

Reply:
The JDT compiler testsuite has > 10.000 tests and is parametrized. This is one of the "complex" test suites I personally know that has more complex structure as "usual" tests and so a real world example that shouldn't be broken.

Reply:
> The JDT compiler testsuite has > 10.000 tests and is parametrized. This is one of the "complex" test suites I personally know that has more complex structure as "usual" tests and so a real world example that shouldn't be broken.

There are tons of errors when I run that test case:

![TestAll_CCE_dialog](https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/24752155/011b6b92-e080-49c8-bbda-53fbd71040f4)

I'll try to update my Eclipse. It might be a problem with how I run the test though, I have imported the test plug-in in a debuggee Eclipse... Maybe something goes wrong that way.

I also see a CME with my change (I'm not sure if its seen without it, but I guess it should be possible with current `master`):

```
java.util.ConcurrentModificationException
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1631)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:575)
	at java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)
	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:616)
	at org.eclipse.jdt.internal.junit.model.TestSuiteElement.getChildren(TestSuiteElement.java:50)
	at org.eclipse.jdt.internal.junit.ui.TestSessionTreeContentProvider.hasChildren(TestSessionTreeContentProvider.java:54)
	at org.eclipse.jface.viewers.AbstractTreeViewer.isExpandable(AbstractTreeViewer.java:2238)
	at org.eclipse.jface.viewers.TreeViewer.isExpandable(TreeViewer.java:557)
	at org.eclipse.jface.viewers.AbstractTreeViewer.isExpandable(AbstractTreeViewer.java:2268)
	at org.eclipse.jface.viewers.AbstractTreeViewer.updatePlus(AbstractTreeViewer.java:2996)
	at org.eclipse.jface.viewers.TreeViewer.updatePlus(TreeViewer.java:805)
	at org.eclipse.jface.viewers.AbstractTreeViewer.createTreeItem(AbstractTreeViewer.java:878)
	at org.eclipse.jface.viewers.AbstractTreeViewer.createChildren(AbstractTreeViewer.java:855)
	at org.eclipse.jface.viewers.TreeViewer.createChildren(TreeViewer.java:615)
	at org.eclipse.jface.viewers.AbstractTreeViewer.createChildren(AbstractTreeViewer.java:803)
	at org.eclipse.jface.viewers.AbstractTreeViewer.internalInitializeTree(AbstractTreeViewer.java:1627)
	at org.eclipse.jface.viewers.TreeViewer.internalInitializeTree(TreeViewer.java:791)
	at org.eclipse.jface.viewers.AbstractTreeViewer.lambda$1(AbstractTreeViewer.java:1612)
	at org.eclipse.jface.viewers.StructuredViewer.preservingSelection(StructuredViewer.java:1397)
	at org.eclipse.jface.viewers.TreeViewer.preservingSelection(TreeViewer.java:366)
	at org.eclipse.jface.viewers.StructuredViewer.preservingSelection(StructuredViewer.java:1358)
	at org.eclipse.jface.viewers.AbstractTreeViewer.inputChanged(AbstractTreeViewer.java:1606)
	at org.eclipse.jface.viewers.ContentViewer.setInput(ContentViewer.java:282)
	at org.eclipse.jface.viewers.StructuredViewer.setInput(StructuredViewer.java:1641)
	at org.eclipse.jdt.internal.junit.ui.TestViewer.processChangesInUI(TestViewer.java:692)
	at org.eclipse.jdt.internal.junit.ui.TestRunnerViewPart.processChangesInUI(TestRunnerViewPart.java:1392)
	at org.eclipse.jdt.internal.junit.ui.TestRunnerViewPart$UpdateUIJob.runInUIThread(TestRunnerViewPart.java:881)
	at org.eclipse.ui.progress.UIJob.lambda$0(UIJob.java:148)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:132)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:5039)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4519)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:342)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:645)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:342)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:552)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:173)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:651)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:588)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1459)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1432)
```

Reply:
Which SDK build are you usimg? The CCE shouldn't happen in master / 4.30 final. Stack trace?
The CME is most likely caused by this PR (the elements are added while getChildern() loops over them.

I don't see any exceptions using latest SDK build.

Reply:
> Which SDK build are you usimg? The CCE shouldn't happen in master / 4.30 final. Stack trace?

Its not logged, but stopping the debugger at the constructor of CCE, I see:

```
"main" #1 prio=6 os_prio=0 cpu=9724.20ms elapsed=330.45s tid=0x00007ffff0028910 nid=0x1160c at breakpoint [0x00007ffff7fb4000]
   java.lang.Thread.State: RUNNABLE
        at java.lang.ClassCastException.<init>(java.base@17.0.6/ClassCastException.java:57)
        at org.eclipse.jdt.internal.junit.ui.TestSessionLabelProviderCustom.getStyledText(TestSessionLabelProviderCustom.java:56)
        at org.eclipse.jface.viewers.DelegatingStyledCellLabelProvider.getStyledText(DelegatingStyledCellLabelProvider.java:209)
        at org.eclipse.jface.viewers.DecoratingStyledCellLabelProvider.getStyledText(DecoratingStyledCellLabelProvider.java:197)
        at org.eclipse.jface.viewers.DelegatingStyledCellLabelProvider.update(DelegatingStyledCellLabelProvider.java:109)
        at org.eclipse.jface.viewers.DecoratingStyledCellLabelProvider.update(DecoratingStyledCellLabelProvider.java:134)
        at org.eclipse.jface.viewers.ViewerColumn.refresh(ViewerColumn.java:146)
        at org.eclipse.jface.viewers.AbstractTreeViewer.doUpdateItem(AbstractTreeViewer.java:983)
        at org.eclipse.jface.viewers.AbstractTreeViewer$UpdateItemSafeRunnable.run(AbstractTreeViewer.java:130)
        at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:47)
        at org.eclipse.ui.internal.JFaceUtil$$Lambda$112/0x0000000800d88c70.run(Unknown Source)
        at org.eclipse.jface.util.SafeRunnable.run(SafeRunnable.java:174)
        at org.eclipse.jface.viewers.AbstractTreeViewer.doUpdateItem(AbstractTreeViewer.java:1058)
        at org.eclipse.jface.viewers.StructuredViewer$UpdateItemSafeRunnable.run(StructuredViewer.java:427)
        at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:47)
        at org.eclipse.ui.internal.JFaceUtil$$Lambda$112/0x0000000800d88c70.run(Unknown Source)
        at org.eclipse.jface.util.SafeRunnable.run(SafeRunnable.java:174)
        at org.eclipse.jface.viewers.StructuredViewer.updateItem(StructuredViewer.java:2116)
        at org.eclipse.jface.viewers.AbstractTreeViewer.createTreeItem(AbstractTreeViewer.java:877)
        at org.eclipse.jface.viewers.AbstractTreeViewer.createChildren(AbstractTreeViewer.java:855)
        at org.eclipse.jface.viewers.TreeViewer.createChildren(TreeViewer.java:615)
        at org.eclipse.jface.viewers.AbstractTreeViewer.createChildren(AbstractTreeViewer.java:803)
        at org.eclipse.jface.viewers.AbstractTreeViewer.internalExpand(AbstractTreeViewer.java:1718)
        at org.eclipse.jface.viewers.AbstractTreeViewer.reveal(AbstractTreeViewer.java:2387)
        at org.eclipse.jdt.internal.junit.ui.TestViewer.autoScrollInUI(TestViewer.java:826)
        at org.eclipse.jdt.internal.junit.ui.TestViewer.processChangesInUI(TestViewer.java:724)
        at org.eclipse.jdt.internal.junit.ui.TestRunnerViewPart.processChangesInUI(TestRunnerViewPart.java:1392)
        at org.eclipse.jdt.internal.junit.ui.TestRunnerViewPart$UpdateUIJob.runInUIThread(TestRunnerViewPart.java:881)
```

> The CME is most likely caused by this PR (the elements are added while getChildern() loops over them.

`getChildren()` iterates over the elements already on `master` though. Maybe its in the new method, that checks if the parent suite has only 1 dynamic child... Though there we access `fChildren` directly...

> I don't see any exceptions using latest SDK build.

I don't see it in my work Eclipse either, maybe its "just" in the workspace I've been using to test things...

Reply:
I can't find `org.eclipse.jdt.internal.junit.ui.TestSessionLabelProviderCustom` neither in my SDK workspace that has almost everything nor in this repository. Where it is coming from?

Reply:
Looks like we have another problem in `org.eclipse.jdt.internal.junit.ui.TestSessionLabelProviderCustom.getStyledText(Object)`... @iloveeclipse how is the CCE dealt with in this case?

```
	@Override
	public StyledString getStyledText(Object element) {
		ITestElement testElement = (ITestElement) element;
```

If I receive an `ExpandableNode`, just ignore it? I guess we'll have this problem in our IDE...

Reply:
> If I receive an `ExpandableNode`, just ignore it?

Yes. But looking at

https://github.com/eclipse-platform/eclipse.platform.ui/blob/9ebaef1f58cb4c84aea3f222b585288629e626e6/bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java#L975C12-L975C12

I wonder why do we have this assumption about "expandable node only in first column".
@raghucssit : can you please look at that?

Reply:
> I can't find `org.eclipse.jdt.internal.junit.ui.TestSessionLabelProviderCustom` neither in my SDK workspace that has almost everything nor in this repository. Where it is coming from?

Its in the custom plug-in we have for Xpect: https://github.com/meysholdt/eclipse_jdt_junit_runners.git